### PR TITLE
Fix non base 10 number parsing

### LINKED ISSUE #525: Failure to parse hexadecimal and octal numbers in YAML
**Discussion?**

No

**Describe the bug**

Dasel 3.3.1 fails to parse hexadecimal and octal numbers in YAML.

**To Reproduce**

```sh
❯ dasel version
v3.3.1
❯ dasel -i yaml -o json <<<'0x10'
dasel: error: error reading input: yaml: unmarshal errors:
                line 1: strconv.Atoi: parsing "0x10": invalid syntax
❯ dasel -i yaml -o json <<<'0o10'
dasel: error: error reading input: yaml: unmarshal errors:
                line 1: strconv.Atoi: parsing "0o10": invalid syntax
```

Note: Though binary numbers and 0-prefixed octal numbers were abolished in YAML 1.2, those interpretation had changed in dasel v3.

```sh
❯ dasel -i yaml -o json <<<'0b10'
dasel: error: error reading input: yaml: unmarshal errors:
                line 1: strconv.Atoi: parsing "0b10": invalid syntax
❯ dasel -i yaml -o json <<<'010'
10
```

**Expected behavior**

dasel parses `0x..` and `0o..` into numbers.

```sh
❯ dasel -i yaml -o json <<<'0x10'
16
❯ dasel -i yaml -o json <<<'0o10'
8
```

dasel 2.8.1 worked as expected.

```sh
❯ ASDF_DASEL_VERSION=2.8.1 dasel -r yaml -w json <<<'0x10'
16
❯ ASDF_DASEL_VERSION=2.8.1 dasel -r yaml -w json <<<'0o10'
8
❯ ASDF_DASEL_VERSION=2.8.1 dasel -r yaml -w json <<<'0b10'
2
❯ ASDF_DASEL_VERSION=2.8.1 dasel -r yaml -w json <<<'010'
8
```

**Desktop (please complete the following information):**

- OS: macOS Tahoe 26.3
- Version: 3.3.1

**Additional context**

- #326

### PR DESCRIPTION
**Pull request by TomWright**
Fixes #525

### CODE REVIEW

File: execution/func_to_int.go [hunk 1]
```diff
@@ -19,7 +19,7 @@ var FuncToInt = NewFunc(
 				return nil, err
 			}
 
-			i, err := strconv.ParseInt(stringValue, 10, 64)
+			i, err := strconv.ParseInt(stringValue, 0, 64)
```
Comment (ccoVeille) [inline execution/func_to_int.go]: I have been there. I would discourage doing this.

YAML spec is not linked to what Go supports with base 0

Add tests for 0x42 0X42 0o42 0b11

File: parsing/yaml/yaml_test.go [hunk 2]
```diff
@@ -204,6 +204,91 @@ name2: Tom
 `,
 	}.run)
 
+	t.Run("base numbers", func(t *testing.T) {
```
Comment (ccoVeille) [inline parsing/yaml/yaml_test.go:207]: Here are some edge cases you could consider

- Explicit plus sign (+) with integers / prefixed bases
The implementation strips a leading sign before prefix detection, but no tests cover the `+` form.
- `+42` -> `42`
- `+0x10` -> `16`
- `+0o10` -> `8`
- `+0b10` -> `2`

- `-42` is tested, but not negative hex/octal/binary.
- `-0x10` -> `-16`
- `-0o10` -> `-8`
- `-0b10` -> `-2`

- Numeric separators `_` 
- `1_000` -> `1000`
- `0xFF_FF` -> `65535`
- `0b1010_1010` -> `170`

Go’s strconv.ParseInt / ParseFloat won’t accept `_` directly, so this needs either normalization or a documented rejection + tests.

File: parsing/toml/toml_reader.go [hunk 3]
```diff
@@ -150,7 +150,7 @@ func (j *tomlReader) readNode(p *unstable.Parser, n *unstable.Node) (string, *mo
 		}
 		return "", model.NewFloatValue(f), nil
 	case unstable.Integer:
-		i64, err := strconv.ParseInt(string(n.Data), 10, 64)
+		i64, err := strconv.ParseInt(string(n.Data), 0, 64)
```
Comment (ccoVeille) [inline parsing/toml/toml_reader.go:153]: I'm not an expert in TOML, but I took a quick look at the spec

- Please add tests for these:

Valid

42, -42, 0x12, 0o7, 0b111, 12_000, -12_000

Valid but as string: "042"

Invalid

-0x12, -0o7, -0b111, 042

All these values are valid with base 0 with ParseInt but not in TOML

Here are also feedbacks provided by an LLM

```
# ==========================================================
# HEXADECIMAL, OCTAL, AND BINARY BASES
# ==========================================================

# --- Valid (Case Insensitivity for Prefixes and Values) ---
hex_lowercase = 0xff12       # Standard hex
hex_uppercase = 0XFF12       # Uppercase prefix and values are VALID

bin_lowercase = 0b1101
bin_uppercase = 0B1101       # Uppercase 'B' is VALID
oct_lowercase = 0o755
oct_uppercase = 0O755        # Uppercase 'O' is VALID

# --- Invalid ---
# Negative signs are NOT supported for non-decimal bases
invalid_hex = -0xFF12        
invalid_oct = -0o755         
invalid_bin = -0b1101        

# ==========================================================
# SCIENTIFIC NOTATION (FLOATS)
# ==========================================================

# --- Valid ---
standard_sci = 1e6           # 1,000,000
positive_exp = 5e+10         # 50,000,000,000
negative_exp = -1.5E-4       # -0.00015 (Uppercase 'E' is VALID)
large_float  = 6_022.140e23  # Avogadro's number

# --- Invalid ---
# Decimal points must always have digits on both sides
invalid_point_1 = .1e2       
invalid_point_2 = 1.e2       

# Underscores are NOT allowed in the exponent part
invalid_underscore = 1e1_0   
```

**Comment by ccoVeille** [comment]
Good additions.

Is there anything else you want to change?

Should I wait before reviewing again?

@tomwright

**Comment by TomWright** [comment]
> Good additions.
> 
> Is there anything else you want to change?
> 
> Should I wait before reviewing again?
> 
> @TomWright

I think that's all I've got for the time being. If you see any major issues please flag them, otherwise I'm happy to get the main issues resolved and follow-up with minor fixes for specific cases at a later date.

It should be ready for you, thank you!
